pkg/agent/bgp/gobgp/gobgp.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gobgp
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"time"
+
+	gobgpapi "github.com/osrg/gobgp/v3/api"
+	"github.com/osrg/gobgp/v3/pkg/server"
+	"google.golang.org/protobuf/types/known/anypb"
+	"k8s.io/utils/net"
+
+	"antrea.io/antrea/pkg/agent/bgp"
+)
+
+const (
+	ipv4AllZero = "0.0.0.0"
+	ipv6AllZero = "::"
+)
+
+type Server struct {
+	server       *server.BgpServer
+	globalConfig *gobgpapi.Global
+}
+
+func NewGoBGPServer(globalConfig *bgp.GlobalConfig) bgp.Interface {
+	return &Server{
+		server: server.NewBgpServer(server.LoggerOption(newGoBGPLogger())),
+		globalConfig: &gobgpapi.Global{
+			Asn:        globalConfig.ASN,
+			RouterId:   globalConfig.RouterID,
+			ListenPort: globalConfig.ListenPort,
+		},
+	}
+}
+
+func (s *Server) Start(ctx context.Context) error {
+	go s.server.Serve()
+	if err := s.server.StartBgp(ctx, &gobgpapi.StartBgpRequest{Global: s.globalConfig}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) Stop(ctx context.Context) error {
+	if err := s.server.StopBgp(ctx, &gobgpapi.StopBgpRequest{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) AddPeer(ctx context.Context, peer bgp.PeerConfig) error {
+	request := &gobgpapi.AddPeerRequest{Peer: toGoBGPPeer(peer)}
+	if err := s.server.AddPeer(ctx, request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) UpdatePeer(ctx context.Context, peer bgp.PeerConfig) error {
+	request := &gobgpapi.UpdatePeerRequest{Peer: toGoBGPPeer(peer)}
+	if _, err := s.server.UpdatePeer(ctx, request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) RemovePeer(ctx context.Context, peer bgp.PeerConfig) error {
+	request := &gobgpapi.DeletePeerRequest{Address: peer.Address}
+	if err := s.server.DeletePeer(ctx, request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) GetPeers(ctx context.Context) ([]bgp.PeerStatus, error) {
+	var peerStatuses []bgp.PeerStatus
+
+	fn := func(peer *gobgpapi.Peer) {
```

Comment:
have you considered extracting this so you can unit test it?

Reply:
Done